tasking/utils.py

```diff
@@ -258,32 +293,81 @@ def get_shapefile(geofile):
     """
     # Takes the inputted geofile(zipfile) and lists all items in it
     name_list = geofile.namelist()
-    # Initializes name variable
-    name = None
-
-    # Check if zipfile has more than 3 files
-    if len(name_list) > 3:
-        # Raise UnnecessaryFiles Exception if files exceed 3
-        raise UnnecessaryFiles()
-    # Check if zipfile has less than the 3 required files
-    elif len(name_list) < 3:
-        # Raise MissingFiles Exeption if it has less than the required files
+
+    if settings.TASKING_CHECK_NUMBER_OF_FILES_IN_SHAPEFILES_DIR:
+        # Check if zipfile has more than 3 files
+        if len(name_list) > 3:
+            # Raise UnnecessaryFiles Exception if files exceed 3
+            raise UnnecessaryFiles()
+        # Check if zipfile has less than the 3 required files
+        elif len(name_list) < 3:
+            # Raise MissingFiles Exception
+            raise MissingFiles()
+
+    needed_files = {}
+
+    for item in name_list:
+        if item.endswith('shp'):
+            needed_files['shp'] = item
+        elif item.endswith('dbf'):
+            needed_files['dbf'] = item
+        elif item.endswith('shx'):
+            needed_files['shx'] = item
+
+    if not needed_files.get('dbf') or not needed_files.get('shx'):
         raise MissingFiles()
-    # Check if zipfile has 3 files only
-    elif len(name_list) == 3:
-        # Iterate through the names of items to find the .shp file
-        for item in name_list:
-            # Split the elements of the name_list in order to have an array
-            # Of filename and extension name
-            arr = item.split('.')
-            # Check if the extension of the file is .shp
-            if arr[1] == 'shp':
-                # Set name to the name of the .shp file
-                name = '.'.join(arr)
-    # Check if name has changed from its initial value
-    if name is None:
-        # Raise ShapeFileNotFound exception if name hasn't changed from initial
+
+    if not needed_files.get('shp'):
         raise ShapeFileNotFound()
-    else:
-        # Return name
-        return name
+
+    return needed_files['shp']
+
+
+def get_polygons(geom_object_list):
+    """
+    Takes a geom object list and returns polygons, runs recursively
+
+    :param geom_object_list: list of geom objects
+
+    :return: list of Polygon objects
+    """
+
+    def _process_multipolygon(multipolygon_obj, results, ignore_invalid):
+        """
+        Process multipolygon object
+        """
+        if settings.TASKING_SHAPEFILE_ALLOW_NESTED_MULTIPOLYGONS:
+            nested_items = get_polygons((x for x in multipolygon_obj))
+            results = results + nested_items
+        else:
+            if ignore_invalid:
+                pass
+            else:
+                results.append(multipolygon_obj)
+
+        return results
+
+    result = []
+    for item in geom_object_list:
+        if settings.TASKING_SHAPEFILE_IGNORE_INVALID_TYPES:
```

Comment:
What are the invalid types? MultuPolygons? Why?

Reply:
Invalid Types are anything other than Polygons - the ones we know about so far are MultiPolygons and Points.

They are invalid because `multipolygon = MultiPolygon(polygons)` fails if `polygons` contains anything other than `Polygon` objects.